### PR TITLE
fix(typescript-estree): replace `fast-glob` with `tinyglobby`

### LINKED ISSUE #10533: Enhancement: replace `fast-glob` with `tinyglobby`
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

`fast-glob` has 17 dependencies with 12 people having publish access to those libraries and takes 505kb after installation
https://npmgraph.js.org/?q=fast-glob
https://pkg-size.dev/fast-glob

`tinyglobby` has only 2 dependencies with 6 people having publish access to those libraries and takes 155kb after installation
https://npmgraph.js.org/?q=tinyglobby
https://pkg-size.dev/tinyglobby

I would be happy to send a PR for this

### Additional Info

This was discussed a bit in https://github.com/typescript-eslint/typescript-eslint/issues/9453, but I thought it would be a good time to revisit as `tinyglobby` usage has grown 30x since that time having been adopted by projects like `vite`, `nx`, and `nuxt`. `tinyglobby` is likely to dedupe in user projects now. E.g. this repo already pulls in `tinyglobby` multiple times (via `nx` and `cspell`). And other projects in the eslint ecosystem like `eslint-import-resolver-typescript` (downloaded 10m times per week) depend on `tinyglobby` already. While `fast-glob` is still downloaded more frequently, 80% of those downloads come from this project, so whatever library this project uses will end up being the most used.

### PR DESCRIPTION
**Pull request by james-pre**
## PR Checklist

- [x] Addresses an existing open issue: fixes #10533
- [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This PR migrates from `fast-glob` to `tinyglobby`. For more information, you can see the issue that this PR fixes: #10533

### CODE REVIEW

**Comment by james-pre** [comment]
The error message for the test is nonsensical to me, as `_tinyglobby` doesn't even exist in the scope:

```

 FAIL  tests/lib/parse.test.ts
  ● Test suite failed to run
    ReferenceError: Cannot access '_tinyglobby' before initialization
      45 |   return {
      46 |     ...tg,
    > 47 |     globSync: jest.fn(tinyglobby.globSync),
         |                       ^
      48 |   };
      49 | });
      50 |
      at tinyglobby (tests/lib/parse.test.ts:47:23)
      at Object.<anonymous> (tests/lib/parse.test.ts:34:61)
```

File: packages/typescript-estree/tests/lib/parse.test.ts [hunk 1]
```diff
@@ -40,11 +40,11 @@ jest.mock('typescript', () => {
...
 
-jest.mock('fast-glob', () => {
-  const fastGlob = jest.requireActual<typeof fastGlobModule>('fast-glob');
+jest.mock('tinyglobby', () => {
+  const tg = jest.requireActual<typeof tinyglobby>('tinyglobby');
   return {
-    ...fastGlob,
-    sync: jest.fn(fastGlob.sync),
+    ...tg,
+    globSync: jest.fn(tinyglobby.globSync),
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/parse.test.ts:47]: Previously the thing being passed to `jest.fn` was from the `jest.requireActual` object. So to keep to that:

```suggestion
    globSync: jest.fn(tg.globSync),
```

**Comment by JoshuaKGoldberg** [comment]
@james-pre I appreciate the enthusiasm for getting this in 🙂. But was there a reason you deleted the PR template? We have it because we like it and it makes sure folks know the steps they're meant to take before sending a PR.

I'm converting this to draft because the issue hasn't been accepted and there are some test failures. Feel free to keep working on it in the meantime, it's great to see it almost fully working! 🚀

File: packages/typescript-estree/tests/lib/parse.test.ts (hunk 1 above)
Comment (james-pre) [reply packages/typescript-estree/tests/lib/parse.test.ts]: Thanks, I totally missed that!

**Comment by james-pre** [comment]
Sorry on the PR check list, I re-added it. Please let me know if there is anything I can do to move this forward.

**Comment by JoshuaKGoldberg** [comment]
Thanks! There are two blockers I think:

* More input from the team on #10533, just in case there's a better alternative strategy
* CI -in particular unit tests- passing on this PR

For issue input, we're just waiting. Given that this is holiday season it might be a business week or two till we get that.

For CI, that's something you could do.

File: packages/typescript-estree/src/parseSettings/resolveProjectList.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import debug from 'debug';
-import { sync as globSync } from 'fast-glob';
 import isGlob from 'is-glob';
+import { globSync } from 'tinyglobby';
```
Comment (benmccann) [inline packages/typescript-estree/src/parseSettings/resolveProjectList.ts:3]: we don't need the `is-glob` package anymore. we can use `isDynamicPattern` instead
```suggestion
import { globSync, isDynamicPattern } from 'tinyglobby';
```

File: packages/typescript-estree/package.json [hunk 3]
```diff
@@ -57,10 +57,10 @@
     "@typescript-eslint/types": "8.18.1",
     "@typescript-eslint/visitor-keys": "8.18.1",
     "debug": "^4.3.4",
-    "fast-glob": "^3.3.2",
     "is-glob": "^4.0.3",
```
Comment (benmccann) [inline packages/typescript-estree/package.json:60]:
```suggestion
```

File: yarn.lock [hunk 4]
```diff
@@ -5879,14 +5879,14 @@ __metadata:
     "@typescript-eslint/types": 8.18.1
     "@typescript-eslint/visitor-keys": 8.18.1
     debug: ^4.3.4
-    fast-glob: ^3.3.2
     glob: "*"
     is-glob: ^4.0.3
```
Comment (benmccann) [inline yarn.lock:5883]: did you regenerate the lockfile? I'd expect this to be removed now

**Comment by JoshuaKGoldberg** [comment]
Per https://github.com/typescript-eslint/typescript-eslint/issues/10533#issuecomment-2560375811 we're going to hold off for a few months on the switch. `tinyglobby` is not stable enough yet for us to use it.

This is a big part of why we ask that folks go by the contributing guidelines. It feels bad to close a PR with good work in it like this one. But you never know when a seemingly straightforward+good issue is actually not ready to be accepted yet.

Thanks! ❤️

**Comment by james-pre** [comment]
@JoshuaKGoldberg All of the tests passed, so I don't believe the instability issues mentioned in the comment are applicable. What do you think?

**Comment by JoshuaKGoldberg** [comment]
(answering in #10533)

**Pull request closed without merging**